cpp/src/jni/orc/jni_wrapper.cpp

```diff
@@ -0,0 +1,287 @@
+    JNIEnv* env, jobject this_obj, jstring file_path) {
+  std::shared_ptr<arrow::io::ReadableFile> in_file;
+
+  std::string path = JStringToCString(env, file_path);
+
+  arrow::Status ret;
+  if (path.find("hdfs://") == 0) {
+    env->ThrowNew(io_exception_class, "hdfs path not support yet.");
+  } else {
+    ret = arrow::io::ReadableFile::Open(path, &in_file);
+  }
+
+  if (ret.ok()) {
+    std::unique_ptr<ORCFileReader> reader;
+
+    ret = ORCFileReader::Open(
+        std::static_pointer_cast<arrow::io::RandomAccessFile>(in_file),
+        arrow::default_memory_pool(), &reader);
+
+    if (!ret.ok()) {
+      env->ThrowNew(io_exception_class, std::string("Failed open file" + path).c_str());
+    }
+
+    return orc_reader_holder_.Insert(std::shared_ptr<ORCFileReader>(reader.release()));
+  }
+
+  return static_cast<jlong>(ret.code()) * -1;
+}
+
+JNIEXPORT void JNICALL Java_org_apache_arrow_adapter_orc_OrcReaderJniWrapper_close(
+    JNIEnv* env, jobject this_obj, jlong id) {
+  orc_reader_holder_.Erase(id);
+}
+
+JNIEXPORT jboolean JNICALL Java_org_apache_arrow_adapter_orc_OrcReaderJniWrapper_seek(
+    JNIEnv* env, jobject this_obj, jlong id, jint row_number) {
+  auto reader = orc_reader_holder_.Lookup(id);
+  return reader->Seek(row_number).ok();
+}
+
+JNIEXPORT jint JNICALL
+Java_org_apache_arrow_adapter_orc_OrcReaderJniWrapper_getNumberOfStripes(JNIEnv* env,
+                                                                         jobject this_obj,
+                                                                         jlong id) {
+  auto reader = orc_reader_holder_.Lookup(id);
+  return reader->NumberOfStripes();
+}
+
+JNIEXPORT jlong JNICALL
+Java_org_apache_arrow_adapter_orc_OrcReaderJniWrapper_nextStripeReader(JNIEnv* env,
+                                                                       jobject this_obj,
+                                                                       jlong id,
+                                                                       jlong batch_size) {
+  auto reader = GetNativeReader(id);
+  std::shared_ptr<RecordBatchReader> stripe_reader;
+  auto status = reader->NextStripeReader(batch_size, &stripe_reader);
+
+  if (!status.ok()) {
+    return static_cast<jlong>(status.code()) * -1;
+  }
+
+  if (!stripe_reader) {
+    return static_cast<jlong>(arrow::StatusCode::Invalid) * -1;
+  }
+
+  return orc_stripe_reader_holder_.Insert(stripe_reader);
+}
+
+JNIEXPORT jbyteArray JNICALL
+Java_org_apache_arrow_adapter_orc_OrcStripeReaderJniWrapper_getSchema(JNIEnv* env,
+                                                                      jclass this_cls,
+                                                                      jlong id) {
+  auto stripe_reader = GetStripeReader(id);
+  auto schema = stripe_reader->schema();
+
+  std::shared_ptr<arrow::Buffer> out;
+  auto status =
+      arrow::ipc::SerializeSchema(*schema, nullptr, arrow::default_memory_pool(), &out);
```

Comment:
see comment above about default_memory_pool.  not necessary for this change but it might be nice to consolidate memory pool references to a single varaible which would make it easier to change.

Reply:
I leave it as default_memory_pool because it's kind of difficult for users on Java side to decide which memory to pick. I will submit a followup PR after this one that focus on integrating memory allocated in native code to the Arrow Allocator counting. I will handle the memory_pool issue there.